cs/src/core/Index/FasterLog/FasterLog.cs

```diff
@@ -0,0 +1,472 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#pragma warning disable 0162
+
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace FASTER.core
+{
+
+    /// <summary>
+    /// FASTER log
+    /// </summary>
+    public class FasterLog : IDisposable
+    {
+        private readonly BlittableAllocator<Empty, byte> allocator;
+        private readonly LightEpoch epoch;
+        private ILogCommitManager logCommitManager;
+        private TaskCompletionSource<long> commitTcs = new TaskCompletionSource<long>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        /// <summary>
+        /// Beginning address of log
+        /// </summary>
+        public long BeginAddress => allocator.BeginAddress;
+
+        /// <summary>
+        /// Tail address of log
+        /// </summary>
+        public long TailAddress => allocator.GetTailAddress();
+
+        /// <summary>
+        /// Log flushed until address
+        /// </summary>
+        public long FlushedUntilAddress => allocator.FlushedUntilAddress;
+
+        /// <summary>
+        /// Log commit until address
+        /// </summary>
+        public long CommittedUntilAddress;
+
+        /// <summary>
+        /// Task notifying commit completions
+        /// </summary>
+        public Task<long> CommitTask => commitTcs.Task;
+
+        /// <summary>
+        /// Create new log instance
+        /// </summary>
+        /// <param name="logSettings"></param>
+        public FasterLog(FasterLogSettings logSettings)
+        {
+            logCommitManager = logSettings.LogCommitManager ?? 
+                new LocalLogCommitManager(logSettings.LogCommitFile ??
+                logSettings.LogDevice.FileName + ".commit");
+
+            epoch = new LightEpoch();
+            CommittedUntilAddress = Constants.kFirstValidAddress;
+            allocator = new BlittableAllocator<Empty, byte>(
+                logSettings.GetLogSettings(), null, 
+                null, epoch, e => Commit(e));
+            allocator.Initialize();
+            Restore();
+        }
+
+        /// <summary>
+        /// Dispose
+        /// </summary>
+        public void Dispose()
```

Comment:
please do a commitTcs.TrySetException here to release any waiting readers/writers when the log is shutdown?

Reply:
done